libs/ml/include/ml/optimisation/optimiser.hpp

```diff
@@ -125,408 +122,6 @@ class Optimiser
-}
-/**
- *
- * @tparam T
- * @tparam C
- * @param new_learning_rate
- */
-template <class T>
-void Optimiser<T>::UpdateLearningRate()
-{
-  switch (learning_rate_param_.mode)
-  {
-  case LearningRateParam<DataType>::LearningRateDecay::EXPONENTIAL:
-  {
-    learning_rate_ = learning_rate_param_.starting_learning_rate *
-                     fetch::math::Pow(learning_rate_param_.exponential_decay_rate,
-                                      static_cast<DataType>(epoch_ + 1));
-    break;
-  }
-  case LearningRateParam<DataType>::LearningRateDecay::LINEAR:
-  {
-    learning_rate_ = learning_rate_param_.starting_learning_rate *
-                     (DataType{1} - learning_rate_param_.linear_decay_rate *
-                                        static_cast<DataType>(cumulative_step_));
-    if (learning_rate_ < learning_rate_param_.ending_learning_rate)
-    {
-      learning_rate_ = learning_rate_param_.ending_learning_rate;
-    }
-    break;
-  }
-  case LearningRateParam<DataType>::LearningRateDecay::NONE:
-  {
-    break;
-  }
-  default:
-  {
-    throw exceptions::InvalidMode("Please specify learning rate schedule method");
-  }
-  }
-}
-
-/**
- * Increments epoch counter
- * @tparam T
- */
-template <class T>
-void Optimiser<T>::IncrementEpochCounter()
-{
-  epoch_++;
-}
-
-/**
- * Increments batch counters
- * @tparam T
- */
-template <class T>
-void Optimiser<T>::IncrementBatchCounters(SizeType batch_size)
-{
-  step_ += batch_size;
-  cumulative_step_ += batch_size;
-}
-
-/**
- * helper method for setting or updating batch size in case of invalid parameter combinations.
- * For example, batch_size cannot be larger than data_size or subset_size.
- * @tparam T
- * @tparam C
- * @param batch_size
- * @param data_size
- * @param subset_size
- * @return
- */
-template <class T>
-typename Optimiser<T>::SizeType Optimiser<T>::UpdateBatchSize(SizeType const &batch_size,
-                                                              SizeType const &data_size,
-                                                              SizeType const &subset_size)
-{
-  SizeType updated_batch_size = batch_size;
-  // If batch_size not specified do full batch
-  if (batch_size == SIZE_NOT_SET)
-  {
-    if (subset_size == SIZE_NOT_SET)
-    {
-      updated_batch_size = data_size;
-    }
-    else
-    {
-      updated_batch_size = subset_size;
-    }
-  }
-  // If user sets smaller subset size than batch size
-  if (batch_size > subset_size)
-  {
-    updated_batch_size = subset_size;
-  }
-  // if batch size is larger than all data
-  if (batch_size > data_size)
-  {
-    updated_batch_size = data_size;
-  }
-  return updated_batch_size;
-}
-
-template <typename T>
-void Optimiser<T>::ResetGradients()
-{
-  this->graph_->ResetGradients();
-}
-
-template <typename T>
-std::shared_ptr<Graph<T>> Optimiser<T>::GetGraph()
-{
-  return graph_;
-}
-
 }  // namespace optimisers
 }  // namespace ml
 
```

Comment:
move the serialisation/deserialisation code out to serializers/ml_types.hpp
this will allow you to not have to include graph_builder.hpp in this file which is probably the most expensive include in the whole library